### ISSUE TITLE
Background Thread for event upload.

### ISSUE THREAD

**Issue by XerTheSquirrel** [iopipe/iopipe-java#95]
Instead of resolving and opening new connections per event sent, instead have a background thread which takes in events and publishes them. It can open and just idle the next connection to be made. Perhaps there could be a slight idle to make sure events are sent, but maybe not waiting too long for it to happen.

**Comment by XerTheSquirrel** [comment]
To reduce latency when an event is to be posted I was thinking of a queue to push events that just keeps pushing whenever there is stuff in the queue to do so. I would need to keep track of the number of executions that are currently running at once, so if there are events to be sent to IOpipe in the queue then the last invocation will just sit there doing nothing.

For non-concurrent executions this would act exactly the same as before and there would be a slight latency increase (should be very negligible).

For concurrent executions this would decrease latency in individual executions as more threads can do work at once without waiting for the server to respond.

The profiler does something similar with the background polling for the remote connection to upload to (it makes it then just waits), but it can do something similar with a push system. The only thing to consider is if profiling information is permitted to be uploaded to the server after the event has been posted.

If a server returns an error because of a bad token or similar, then it is never an error anyway and just prints to the log. So this will not have any effect on that because it should always be non-fatal anyway.

**Comment by XerTheSquirrel** [comment]
I will need to add two test cases for concurrency (which might need an adjustment in the test system):

 * Invoking things slowly one at a time.
 * As many concurrent events all at the same time. Sort of load testing but just a way to ensure that things are still happening okay with lots of events happening.

There should be no timeout trigger.

**Comment by XerTheSquirrel** [comment]
Also the profiler plugin should also add to the latency for every event since it wants to send the ZIP it generates to S3. So by sharing this framework with the profiler that can get the latency reducing bonuses too.

**Comment by kolanos** [comment]
@XerTheSquirrel Have you seen this in the wild? (i.e. on AWS) My understanding (and this is the way the Python and Go runtimes work) is that there can only be a single invocation running per process. Does the Java runtime behave differently here?

**Comment by XerTheSquirrel** [comment]
@kolanos Yes you are correct, only a single invocation is executed per process. I thought I saw concurrent invocation in the wild and made some assumptions which have turned out to be incorrect. So the PR basically just does the same to before, just that the overflow will not occur since there will never be more than one concurrent invocation.
